src/main/scala/org/broadinstitute/dsde/workbench/leonardo/dao/GoogleDataprocDAO.scala

```diff
@@ -9,59 +10,194 @@ import com.google.api.client.auth.oauth2.Credential
+case class FirewallRuleInsertionException(googleProject: GoogleProject, firewallRule: String) extends LeoException(s"Unable to insert new firewall rule $googleProject/$firewallRule", StatusCodes.InternalServerError)
+
 
-class GoogleDataprocDAO(protected val dataprocConfig: DataprocConfig)(implicit val system: ActorSystem, val executionContext: ExecutionContext)
+class GoogleDataprocDAO(protected val dc: DataprocConfig)(implicit val system: ActorSystem, val executionContext: ExecutionContext)
   extends DataprocDAO with GoogleUtilities {
 
   private val httpTransport = GoogleNetHttpTransport.newTrustedTransport
   private val jsonFactory = JacksonFactory.getDefaultInstance
-  private val cloudPlatformScopes = List(PubsubScopes.CLOUD_PLATFORM)
+  private val cloudPlatformScopes = List(ComputeScopes.CLOUD_PLATFORM)
+  private val storageScopes = Seq(StorageScopes.DEVSTORAGE_FULL_CONTROL, ComputeScopes.COMPUTE, PlusScopes.USERINFO_EMAIL, PlusScopes.USERINFO_PROFILE)
+  private val vmScopes = List(ComputeScopes.COMPUTE, ComputeScopes.CLOUD_PLATFORM)
+  private val serviceAccountPemFile = new File(dc.configFolderPath, dc.serviceAccountPemName)
 
-  private def getDataProcServiceAccountCredential: Credential = {
+  private lazy val getDataProcServiceAccountCredential: Credential = {
     new GoogleCredential.Builder()
       .setTransport(httpTransport)
       .setJsonFactory(jsonFactory)
-      .setServiceAccountId(dataprocConfig.serviceAccount)
+      .setServiceAccountId(dc.serviceAccount)
       .setServiceAccountScopes(cloudPlatformScopes.asJava)
-      .setServiceAccountPrivateKeyFromPemFile(new java.io.File(dataprocConfig.serviceAccountPemPath))
+      .setServiceAccountPrivateKeyFromPemFile(serviceAccountPemFile)
+      .build()
+  }
+
+  private lazy val getBucketServiceAccountCredential: Credential = {
+    new GoogleCredential.Builder()
+      .setTransport(httpTransport)
+      .setJsonFactory(jsonFactory)
+      .setServiceAccountId(dc.serviceAccount)
+      .setServiceAccountScopes(storageScopes.asJava) // grant bucket-creation powers
+      .setServiceAccountPrivateKeyFromPemFile(serviceAccountPemFile)
+      .build()
+  }
+
+  private lazy val getVmServiceAccountCredential: Credential = {
+    new GoogleCredential.Builder()
+      .setTransport(httpTransport)
+      .setJsonFactory(jsonFactory)
+      .setServiceAccountId(dc.serviceAccount)
+      .setServiceAccountScopes(vmScopes.asJava)
+      .setServiceAccountPrivateKeyFromPemFile(serviceAccountPemFile)
       .build()
   }
 
+  def createCluster(googleProject: GoogleProject, clusterName: String, clusterRequest: ClusterRequest)(implicit executionContext: ExecutionContext): Future[ClusterResponse] = {
+    val bucketName = s"${clusterName}-${UUID.randomUUID.toString}"
+
+    updateFirewallRules(googleProject)  //when should this rule be created? Before or after cluster creation?
```

Comment:
did you get an answer to this question? does it matter, if it's applied to the project and not the VM?